File: jupyter2/problems/minc/minc/ml/orig/lib/parse.ml
open Ast

let ( let* ) = Option.bind

(* report an error at the point it occurs and return None *)
let fail msg =
  let () = Printf.eprintf "parse error: %s\n" msg in
  None

type state = {
  toks : Lex.token array;
  mutable pos : int;
}

let mk toks = { toks = Array.of_list toks; pos = 0 }

let peek st = st.toks.(st.pos)

let advance st =
  let t = st.toks.(st.pos) in
  if st.pos < Array.length st.toks - 1 then st.pos <- st.pos + 1;
  t

let expect st tok =
  let t = peek st in
  if t = tok then
    let _ = advance st in
    Some ()
  else
    fail (Printf.sprintf "expected %s but got %s"
            (Lex.to_string tok) (Lex.to_string t))

(* expr = equality_expr "=" expr | equality_expr *)
let rec parse_expr st =
  let* lhs = parse_equality st in
  match peek st with
  | Lex.ASSIGN ->
    let _ = advance st in
    let* rhs = parse_expr st in
    Some (Assign (lhs, rhs))
  | _ -> Some lhs

(* equality_expr = cmp_expr { ("=="|"!=") cmp_expr }* *)
and parse_equality st =
  let rec loop lhs =
    match peek st with
    | Lex.EQ ->
      let _ = advance st in
      let* rhs = parse_cmp st in
      loop (Binary (Eq, lhs, rhs))
    | Lex.NE ->
      let _ = advance st in
      let* rhs = parse_cmp st in
      loop (Binary (Ne, lhs, rhs))
    | _ -> Some lhs
  in
  let* first = parse_cmp st in
  loop first

(* cmp_expr = additive_expr { ("<="|">="|"<"|">") additive_expr }* *)
and parse_cmp st =
  let rec loop lhs =
    match peek st with
    | Lex.LT ->
      let _ = advance st in
      let* rhs = parse_additive st in
      loop (Binary (Lt, lhs, rhs))
    | Lex.GT ->
      let _ = advance st in
      let* rhs = parse_additive st in
      loop (Binary (Gt, lhs, rhs))
    | Lex.LE ->
      let _ = advance st in
      let* rhs = parse_additive st in
      loop (Binary (Le, lhs, rhs))
    | Lex.GE ->
      let _ = advance st in
      let* rhs = parse_additive st in
      loop (Binary (Ge, lhs, rhs))
    | _ -> Some lhs
  in
  let* first = parse_additive st in
  loop first

(* additive_expr = multiplicative_expr { ("+"|"-") multiplicative_expr }* *)
and parse_additive st =
  let rec loop lhs =
    match peek st with
    | Lex.PLUS ->
      let _ = advance st in
      let* rhs = parse_mul st in
      loop (Binary (Add, lhs, rhs))
    | Lex.MINUS ->
      let _ = advance st in
      let* rhs = parse_mul st in
      loop (Binary (Sub, lhs, rhs))
    | _ -> Some lhs
  in
  let* first = parse_mul st in
  loop first

(* multiplicative_expr = unary_expr { ("*"|"/"|"%") unary_expr }* *)
and parse_mul st =
  let rec loop lhs =
    match peek st with
    | Lex.STAR ->
      let _ = advance st in
      let* rhs = parse_unary st in
      loop (Binary (Mul, lhs, rhs))
    | Lex.SLASH ->
      let _ = advance st in
      let* rhs = parse_unary st in
      loop (Binary (Div, lhs, rhs))
    | Lex.PERCENT ->
      let _ = advance st in
      let* rhs = parse_unary st in
      loop (Binary (Mod, lhs, rhs))
    | _ -> Some lhs
  in
  let* first = parse_unary st in
  loop first

(* unary_expr = number | identifier "(" arg_list ")" | identifier | "(" expr ")" | ("+"|"-"|"!"|"~") unary_expr *)
and parse_unary st =
  match peek st with
  | Lex.NUM n ->
    let _ = advance st in
    Some (Int n)
  | Lex.PLUS ->
    let _ = advance st in
    let* e = parse_unary st in
    Some (Unary (Pos, e))
  | Lex.MINUS ->
    let _ = advance st in
    let* e = parse_unary st in
    Some (Unary (Neg, e))
  | Lex.NOT ->
    let _ = advance st in
    let* e = parse_unary st in
    Some (Unary (LNot, e))
  | Lex.TILDE ->
    let _ = advance st in
    let* e = parse_unary st in
    Some (Unary (BNot, e))
  | Lex.LPAREN ->
    let _ = advance st in
    let* e = parse_expr st in
    let* () = expect st Lex.RPAREN in
    Some e
  | Lex.ID name ->
    let _ = advance st in
    (match peek st with
     | Lex.LPAREN ->
       let _ = advance st in
       let* args = parse_arg_list st in
       let* () = expect st Lex.RPAREN in
       Some (Call (name, args))
     | _ -> Some (Var name))
  | t ->
    fail (Printf.sprintf "unexpected token %s in expression" (Lex.to_string t))

(* arg_list = expr { "," expr }* | {} *)
and parse_arg_list st =
  match peek st with
  | Lex.RPAREN -> Some []
  | _ ->
    let rec loop acc =
      let* e = parse_expr st in
      match peek st with
      | Lex.COMMA ->
        let _ = advance st in
        loop (e :: acc)
      | _ -> Some (List.rev (e :: acc))
    in
    loop []

(* type_expr = "long" *)
let parse_type st =
  let* () = expect st Lex.LONG in
  Some TyLong

(* identifier *)
let parse_ident st =
  match advance st with
  | Lex.ID name -> Some name
  | t ->
    fail (Printf.sprintf "expected identifier but got %s" (Lex.to_string t))

(* stmt = ";" | "continue" ";" | "break" ";" | "return" expr ";" | compound_stmt | if_stmt | while_stmt | expr ";" *)
let rec parse_stmt st =
  match peek st with
  | Lex.SEMI ->
    let _ = advance st in
    Some Empty
  | Lex.CONTINUE ->
    let _ = advance st in
    let* () = expect st Lex.SEMI in
    Some Continue
  | Lex.BREAK ->
    let _ = advance st in
    let* () = expect st Lex.SEMI in
    Some Break
  | Lex.RETURN ->
    let _ = advance st in
    let* e = parse_expr st in
    let* () = expect st Lex.SEMI in
    Some (Return e)
  | Lex.LBRACE -> parse_compound st
  | Lex.IF -> parse_if st
  | _ ->
    let* e = parse_expr st in
    let* () = expect st Lex.SEMI in
    Some (Expr e)

(* compound_stmt = "{" {var_decl}* {stmt}* "}" *)
and parse_compound st =
  let* () = expect st Lex.LBRACE in
  let rec decls acc =
    match peek st with
    | Lex.LONG ->
      let* ty = parse_type st in
      let* name = parse_ident st in
      let* () = expect st Lex.SEMI in
      decls ({ vd_type = ty; vd_name = name } :: acc)
    | _ -> Some (List.rev acc)
  in
  let* ds = decls [] in
  let rec stmts acc =
    match peek st with
    | Lex.RBRACE -> Some (List.rev acc)
    | _ ->
      let* s = parse_stmt st in
      stmts (s :: acc)
  in
  let* ss = stmts [] in
  let* () = expect st Lex.RBRACE in
  Some (Compound (ds, ss))

(* if_stmt = "if" "(" expr ")" stmt [ "else" stmt ] *)
and parse_if st =
  let* () = expect st Lex.IF in
  let* () = expect st Lex.LPAREN in
  let* cond = parse_expr st in
  let* () = expect st Lex.RPAREN in
  let* then_s = parse_stmt st in
  match peek st with
  | Lex.ELSE ->
    let _ = advance st in
    let* else_s = parse_stmt st in
    Some (If (cond, then_s, Some else_s))
  | _ -> Some (If (cond, then_s, None))


(* parameter = type_expr identifier *)
let parse_param st =
  let* ty = parse_type st in
  let* name = parse_ident st in
  Some { p_type = ty; p_name = name }

(* parameter_list = parameter { "," parameter }* | {} *)
let parse_param_list st =
  match peek st with
  | Lex.RPAREN -> Some []
  | _ ->
    let rec loop acc =
      let* p = parse_param st in
      match peek st with
      | Lex.COMMA ->
        let _ = advance st in
        loop (p :: acc)
      | _ -> Some (List.rev (p :: acc))
    in
    loop []

(* fun_definition = type_expr identifier "(" parameter_list ")" compound_stmt *)
let parse_definition st =
  let* ret = parse_type st in
  let* name = parse_ident st in
  let* () = expect st Lex.LPAREN in
  let* params = parse_param_list st in
  let* () = expect st Lex.RPAREN in
  let* body = parse_compound st in
  Some (FunDef { f_ret = ret; f_name = name; f_params = params; f_body = body })

(* program = {definition}* *)
let parse_program st =
  let rec loop acc =
    match peek st with
    | Lex.EOF -> Some (List.rev acc)
    | _ ->
      let* d = parse_definition st in
      loop (d :: acc)
  in
  loop []

(* start = program $ *)
let parse (toks : Lex.token list) : program option =
  let st = mk toks in
  let* prog = parse_program st in
  match peek st with
  | Lex.EOF -> Some prog
  | t ->
    fail (Printf.sprintf "trailing tokens, starting at %s" (Lex.to_string t))

let parse_string (s : string) : program option =
  let* toks = Lex.tokenize s in
  parse toks

File: jupyter2/problems/minc/minc/rs/minc/src/parse.rs
use crate::ast::*;
use crate::lex::{tokenize, Tok, tok_to_string};

pub struct TokenStream {
    toks: Vec<Tok>,
    pos: usize,
}

fn peek(t: &TokenStream) -> &Tok {
    &t.toks[t.pos]
}

fn advance(t: &mut TokenStream) -> Tok {
    let tok = t.toks[t.pos].clone();
    if t.pos < t.toks.len() - 1 {
        t.pos += 1;
    }
    tok
}

fn expect(t: &mut TokenStream, k: &Tok) -> Result<(), String> {
    let tok = peek(t);
    if tok != k {
        return Err(format!("parse error: expected {} but got {}", tok_to_string(k), tok_to_string(tok)));
    }
    advance(t);
    Ok(())
}

// expr = equality_expr "=" expr | equality_expr
fn parse_expr(t: &mut TokenStream) -> Result<Expr, String> {
    let lhs = parse_equality(t)?;
    if *peek(t) == Tok::Assign {
        advance(t);
        let rhs = parse_expr(t)?;
        return Ok(Expr::Assign(Box::new(lhs), Box::new(rhs)));
    }
    Ok(lhs)
}

// equality_expr = cmp_expr { ("=="|"!=") cmp_expr }*
fn parse_equality(t: &mut TokenStream) -> Result<Expr, String> {
    let mut lhs = parse_cmp(t)?;
    loop {
        let op = match peek(t) {
            Tok::Eq => BinaryOp::Eq,
            Tok::Ne => BinaryOp::Ne,
            _ => return Ok(lhs),
        };
        advance(t);
        let rhs = parse_cmp(t)?;
        lhs = Expr::Binary(op, Box::new(lhs), Box::new(rhs));
    }
}

// cmp_expr = additive_expr { ("<="|">="|"<"|">") additive_expr }*
fn parse_cmp(t: &mut TokenStream) -> Result<Expr, String> {
    let mut lhs = parse_additive(t)?;
    loop {
        let op = match peek(t) {
            Tok::Lt => BinaryOp::Lt,
            Tok::Gt => BinaryOp::Gt,
            Tok::Le => BinaryOp::Le,
            Tok::Ge => BinaryOp::Ge,
            _ => return Ok(lhs),
        };
        advance(t);
        let rhs = parse_additive(t)?;
        lhs = Expr::Binary(op, Box::new(lhs), Box::new(rhs));
    }
}

// additive_expr = multiplicative_expr { ("+"|"-") multiplicative_expr }*
fn parse_additive(t: &mut TokenStream) -> Result<Expr, String> {
    let mut lhs = parse_mul(t)?;
    loop {
        let op = match peek(t) {
            Tok::Plus => BinaryOp::Add,
            Tok::Minus => BinaryOp::Sub,
            _ => return Ok(lhs),
        };
        advance(t);
        let rhs = parse_mul(t)?;
        lhs = Expr::Binary(op, Box::new(lhs), Box::new(rhs));
    }
}

// multiplicative_expr = unary_expr { ("*"|"/"|"%") unary_expr }*
fn parse_mul(t: &mut TokenStream) -> Result<Expr, String> {
    let mut lhs = parse_unary(t)?;
    loop {
        let op = match peek(t) {
            Tok::Star => BinaryOp::Mul,
            Tok::Slash => BinaryOp::Div,
            Tok::Percent => BinaryOp::Mod,
            _ => return Ok(lhs),
        };
        advance(t);
        let rhs = parse_unary(t)?;
        lhs = Expr::Binary(op, Box::new(lhs), Box::new(rhs));
    }
}

// unary_expr = number | identifier "(" arg_list ")" | identifier | "(" expr ")" | ("+"|"-"|"!"|"~") unary_expr
fn parse_unary(t: &mut TokenStream) -> Result<Expr, String> {
    let tok = peek(t).clone();
    match tok {
        Tok::Num(v) => {
            advance(t);
            Ok(Expr::IntLit(v))
        }
        Tok::Plus => {
            advance(t);
            Ok(Expr::Unary(UnaryOp::Pos, Box::new(parse_unary(t)?)))
        }
        Tok::Minus => {
            advance(t);
            Ok(Expr::Unary(UnaryOp::Neg, Box::new(parse_unary(t)?)))
        }
        Tok::Not => {
            advance(t);
            Ok(Expr::Unary(UnaryOp::LNot, Box::new(parse_unary(t)?)))
        }
        Tok::Tilde => {
            advance(t);
            Ok(Expr::Unary(UnaryOp::BNot, Box::new(parse_unary(t)?)))
        }
        Tok::LParen => {
            advance(t);
            let e = parse_expr(t)?;
            expect(t, &Tok::RParen)?;
            Ok(e)
        }
        Tok::Id(name) => {
            advance(t);
            if *peek(t) == Tok::LParen {
                advance(t);
                let args = parse_arg_list(t)?;
                expect(t, &Tok::RParen)?;
                Ok(Expr::Call(name, args))
            } else {
                Ok(Expr::VarRef(name))
            }
        }
        _ => Err(format!("parse error: unexpected token {} in expression", tok_to_string(&tok))),
    }
}

// arg_list = [ expr { "," expr }* ]
fn parse_arg_list(t: &mut TokenStream) -> Result<Vec<Expr>, String> {
    if *peek(t) == Tok::RParen {
        return Ok(Vec::new());
    }
    let mut args = vec![parse_expr(t)?];
    while *peek(t) == Tok::Comma {
        advance(t);
        args.push(parse_expr(t)?);
    }
    Ok(args)
}

// type_expr = "long"
fn parse_type(t: &mut TokenStream) -> Result<TypeExpr, String> {
    expect(t, &Tok::Long)?;
    Ok(TypeExpr::Long)
}

fn parse_ident(t: &mut TokenStream) -> Result<String, String> {
    let tok = advance(t);
    match tok {
        Tok::Id(s) => Ok(s),
        _ => Err(format!("parse error: expected identifier but got {}", tok_to_string(&tok))),
    }
}

// stmt = ";" | "continue" ";" | "break" ";" | "return" expr ";" | compound_stmt | if_stmt | while_stmt | expr ";"
fn parse_stmt(t: &mut TokenStream) -> Result<Stmt, String> {
    match peek(t) {
        Tok::Semi => {
            advance(t);
            Ok(Stmt::Empty)
        }
        Tok::Continue => {
            advance(t);
            expect(t, &Tok::Semi)?;
            Ok(Stmt::Continue)
        }
        Tok::Break => {
            advance(t);
            expect(t, &Tok::Semi)?;
            Ok(Stmt::Break)
        }
        Tok::Return => {
            advance(t);
            let e = parse_expr(t)?;
            expect(t, &Tok::Semi)?;
            Ok(Stmt::Return(e))
        }
        Tok::LBrace => Ok(Stmt::Compound(parse_compound(t)?)),
        Tok::If => parse_if(t),
        _ => {
            let e = parse_expr(t)?;
            expect(t, &Tok::Semi)?;
            Ok(Stmt::ExprStmt(e))
        }
    }
}

// compound_stmt = "{" {var_decl}* {stmt}* "}"
fn parse_compound(t: &mut TokenStream) -> Result<Compound, String> {
    expect(t, &Tok::LBrace)?;
    let mut decls = Vec::new();
    while *peek(t) == Tok::Long {
        let ty = parse_type(t)?;
        let name = parse_ident(t)?;
        expect(t, &Tok::Semi)?;
        decls.push(VarDecl { ty, name });
    }
    let mut stmts = Vec::new();
    while *peek(t) != Tok::RBrace {
        stmts.push(parse_stmt(t)?);
    }
    expect(t, &Tok::RBrace)?;
    Ok(Compound { decls, stmts })
}

// if_stmt = "if" "(" expr ")" stmt [ "else" stmt ]
fn parse_if(t: &mut TokenStream) -> Result<Stmt, String> {
    expect(t, &Tok::If)?;
    expect(t, &Tok::LParen)?;
    let cond = parse_expr(t)?;
    expect(t, &Tok::RParen)?;
    let then = parse_stmt(t)?;
    let els = if *peek(t) == Tok::Else {
        advance(t);
        Some(Box::new(parse_stmt(t)?))
    } else {
        None
    };
    Ok(Stmt::If(cond, Box::new(then), els))
}


// parameter = type_expr identifier
fn parse_param(t: &mut TokenStream) -> Result<Param, String> {
    let ty = parse_type(t)?;
    let name = parse_ident(t)?;
    Ok(Param { ty, name })
}

// parameter_list = [ parameter { "," parameter }* ]
fn parse_param_list(t: &mut TokenStream) -> Result<Vec<Param>, String> {
    if *peek(t) == Tok::RParen {
        return Ok(Vec::new());
    }
    let mut params = vec![parse_param(t)?];
    while *peek(t) == Tok::Comma {
        advance(t);
        params.push(parse_param(t)?);
    }
    Ok(params)
}

// fun_definition = type_expr identifier "(" parameter_list ")" compound_stmt
fn parse_fun_def(t: &mut TokenStream) -> Result<FunDef, String> {
    let ret_type = parse_type(t)?;
    let name = parse_ident(t)?;
    expect(t, &Tok::LParen)?;
    let params = parse_param_list(t)?;
    expect(t, &Tok::RParen)?;
    let body = parse_compound(t)?;
    Ok(FunDef {
        ret_type,
        name,
        params,
        body,
    })
}

// program = {definition}* $
fn parse_program(t: &mut TokenStream) -> Result<Program, String> {
    let mut funcs = Vec::new();
    while *peek(t) != Tok::Eof {
        funcs.push(parse_fun_def(t)?);
    }
    Ok(Program { funcs })
}

pub fn parse(toks: Vec<Tok>) -> Result<Program, String> {
    let mut t = TokenStream { toks, pos: 0 };
    let prog = parse_program(&mut t)?;
    if *peek(&t) != Tok::Eof {
        return Err(format!(
            "parse error: trailing tokens, starting at {}",
            tok_to_string(peek(&t))
        ));
    }
    Ok(prog)
}

pub fn parse_string(s: &str) -> Result<Program, String> {
    let toks = tokenize(s)?;
    parse(toks)
}
